Dependency Visualization using Dot

janispeyer opened the pull request:
* Added `warpforge-visualization` that outputs dot code for dependency visualization to stdout.
* Added command for graph visualization to `warpforge-cli`
* Also some minor improvements using clippy.
* Changed some modules to modern rust module layout.

janispeyer committed:
Minor improvements using clippy

janispeyer committed:
Added plots to warpforge-api

Also some refactorings to warpforge-api:
* Converted to the modern module style: <mod_name>/mod.rs => <mod_name>.rs
* Moved dependencies which were only used in #[cfg(test)] to "dev-dependencies"
  (Meaning we have less dependencies in our builds now.)
* Added a test_common module, which contains functions used accross
  testing code.

janispeyer committed:
Worked on some simplistic visualization of build tree dependencies

janispeyer committed:
Converted warpforge-cli to modern rust module layout

janispeyer committed:
Some documentation for visualisation code and added command to cli project

warpfork commented on warpforge-visualize/src/lib.rs:
Okay this is the only part of this I wouldn't instantly merge ^^

We should fix the upstream data of course, but also I'd say this should probably run off the thing in `~/.warphome/catalogs/warpsys` (or perhaps any other single catalog there, if there's only one, although that's not really a behavioral convention we've standardized on yet) by default.

janispeyer replied:
Changed the URL to warptools in [7b67f10](https://github.com/warptools/rustwarp/pull/2/commits/7b67f103b7740f5c2d30891d77231f34027294df)
And fixed upstream json files: https://github.com/warptools/warpsys/pull/3

Addind other sources would also be very useful in the future for sure.

warpfork commented on warpforge-visualize/src/lib.rs:
I can see this being handy :D

I think you could pretty much `s/warpsys_//g` out of the function names across the whole area, though.  This is all data layout that's general to all catalogs, and warpsys is just an instance of a catalog.

warpfork replied:
I think it would be great to have a catalog reading module that contains functions to do this (and other kinds of lookups) on either a filesystem-backed implementation or a url-backed implementation.

Not sure how far you want to drive towards that right now in the same PR, though.  I'd be fine seeing this code land now and move around later as we figure it out more.

janispeyer replied:
I think it could be nice to have some functions/api like that in the `warpforge-api` project. But I intentionally only did prototypy code here. For example I would do parallel download with something like tokio if I had to port this to production-like code.

warpfork commented on warpforge-api/src/formula.rs:
```diff
@@ -0,0 +1,41 @@
+pub struct ActionScript {
+	pub interpreter: String,
+	pub contents: Vec<String>,
+	#[serde(default, skip_serializing_if = "<&bool>::not")]
+	pub network: bool,
```
I think this is fine (and probably what golang code is doing), but I wonder if these should be either an `Option<bool>`, or also have a check that forbids deserializing a false.  I'm always a little nervous with things that may implicitly shift serial data during a roundtripping.

janispeyer replied:
Changed to `Option<bool>` in https://github.com/janispeyer/rustwarp/commit/33886870d2ff1cc2d17f32fa79334e47d9aa831e

warpfork commented on warpforge-api/src/formula.rs:
```diff
@@ -0,0 +1,41 @@
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum Action {
```
I think you can blast away the `compute.rs` file now, as you've now here finished what it was supposed to be.  ^^

janispeyer replied:
Blasted away (https://github.com/janispeyer/rustwarp/commit/4049a160e60357ceda6ab319d828343c91ad3d82)

warpfork reviewed (approved):
This is super excellent and brings many things much forward :D

janispeyer committed:
Changed warpsys url to warptools organisation

Previously used a fork becuase there were invalid json files in the
upstream repository.

Resolves https://github.com/warptools/rustwarp/pull/2#discussion_r1439484173_

janispeyer committed:
Changed type of optional `network` fields from `bool` to `Option<bool>`

Resolves https://github.com/warptools/rustwarp/pull/2#discussion_r1439491618

janispeyer committed:
"Blasted away" `compute.rs` as requested by @warpfork

Resolves https://github.com/warptools/rustwarp/pull/2#discussion_r1439492039

warpfork commented:
I think I'm good with merging this.

Future work would still be to get the grapher to peek at the local filesystem catalogs instead of the URL (and work on flags that support both, etc).

But I want the rest of the API stuff to land on the main branch enough that I'd merge now :D
